File: atoci/Arrays & Hashing/cpp/groupAnagrams.cpp
#include<iostream>
#include<unordered_map>
#include<vector>
#include<string>
#include<algorithm> 

using namespace std; 

vector<vector<string>> groupAnagrams(vector<string>& strs) {
    unordered_map<string, vector<string>> mymap; 

    for (const auto& word : strs) {
        const string sorted_word = word; 
        std::sort(sorted_word.begin(), sorted_word.end());           
        mymap[sorted_word].push_back(word);
    }

    vector<vector<string>> result; 
    
    for (const auto& vec : mymap) {
        result.push_back(vec.second);
    }

    return result;
}



int main() {

    vector<string> input_a = {"eat","tea","tan","ate","nat","bat"};
    vector<string> input_b = {"a"};
    vector<string> input_c = {"",""};
    vector<string> input_d = {"aba", "aba","baa"};
    vector<string> input_e = {"tea","","eat","","tea",""}; 


    vector<vector<string>> output = groupAnagrams(input_a);

    for (const auto& group : output) {
        for (const auto& str : group) {
            std::cout << str << " ";
        }
        std::cout << std::endl;
    }



    return 0;
}

File: atoci/Arrays & Hashing/typescript/groupAnagrams.ts
function groupAnagrams(words : string[]): string[][] {
  const mymap: {[key:string]: string[]} = {};
  const results: string[][] = [];

  for (const word of words) {
    let key = word.split('').sort().join('');
    
    if (!mymap[key]) {
      mymap[key] = []; 
    }

    mymap[key].push(word)
  }

  for (const key in mymap) {
    results.push(mymap[key])
  }

  return results 
};


const input_a = ["eat","tea","tan","ate","nat","bat"]
const input_b = ["a"]
const input_c = ["",""]
const input_d = ["aba", "aba","baa"]
const input_e = ["tea","","eat","","tea",""]


console.log(groupAnagrams(input_a))
console.log(groupAnagrams(input_b))
console.log(groupAnagrams(input_c))
console.log(groupAnagrams(input_d))
console.log(groupAnagrams(input_e))